### PR TITLE
Fix checkout order total and payment fees calculation

### LINKED ISSUE #5574: [Admin][Order] "Balance due" state after payments with a transaction fee
## Description

The price shown in backoffice orders does not include transaction fees, for the respective payment methods. After payment is made, there is always an amount still to be charged and the order is in a "balance due" state. Attempting to pay for the open amount leads do an additional charge of the fee.

## Expected Behavior

The transaction fee should be included in the total price: it should be updated according to the choices made on Payments section allowing to pay the full price (fee included).

## Actual Behaviour

The transaction fee is not included. Paying for orders with transaction fee always renders the order to be in a  "balance due" state.

## Steps to Reproduce

Admin:
1. Set up a payment method with a fee.
2. Start placing an order in the backoffice, inserting items and customer details.
3. Proceed to Payments and check the total amount to be paid.
4. Choose the previously set payment method and perform the payment. 
5. Check the state of the order: it should be in "Balance due" with an amount corresponding to the value of the transaction fee.
6. Attempt to pay this open amount with that payment method - an additional fee will be charged...

## Animated Gif/Screenshot

Before payment:

After payment:

## Workaround

If you know, in advance, what the value of the transaction fee is, then you can add this value manually, as a "prediction" of what the total cost will be. This allows to pay for the total amount with one payment.

## Severity

bug-s4: it's annoying, but you can use it

## Your Environment

* Version used: v2.10.0
* Browser name and version: Firefox 75
* Operating System and version (desktop or mobile): Desktop Ubuntu 18.04

### LINKED ISSUE #12512: Update and recalculate fees button not working for payment fees after order adjusted in backoffice
## Description
The order total is not being updated with Payment fees that are % based after adding to an order in the backoffice, even after clicking the 'Update and recalculate fees' button.   

## Expected Behavior
Customer makes and order and choose a payment option with a % fee.  A transaction fee is added to the order. In the backend a product is added to the order, order total changes, 'update and recalcuate fees' button is clicked.  It would be expected that this would result in the transaction fee increasing to reflect the new order total.   

## Actual Behaviour
'Update and recalcuate fees' button is clicked, and the transaction fee is not changed. Thus the order total does not correctly reflect the payment fees that should be attached to the order.  

## Steps to Reproduce
1. Create a payment fee with a % fee.
2. Create a customer order selecting that payment fee.
3. Add a product to that order in the backend.
4. Click the 'Update and recalculate fees' button.
5. Note that the transaction fee doesn't not change to reflect the payment fees that should be charged on the increased order total.

## Animated Gif/Screenshot
Screenshot 2024-05-22 at 6 19 08 AM

https://github.com/openfoodfoundation/openfoodnetwork/assets/88801240/e13aeaaa-8483-41c6-8012-24e3541d4b11

## Workaround
Need to manually add an adjustment to the order to reflect the missing payment fees.  Many hub/shop managers may not notice the issue and thus fees would be getting charged incorrectly.  

## Severity
bug-s3: a feature is broken but there is a workaround

https://github.com/openfoodfoundation/openfoodnetwork/wiki/Bug-severity
-->

## Your Environment

* Version used: openfoodnetwork.ca
* Browser name and version: Firefox 126.0 (64-bit)
* Operating System and version (desktop or mobile): Mac 13.6.6 (22G630) desktop

### PR DESCRIPTION
**Pull request by rioug**
#### What? Why?

- Closes #5574 
- Closes #12512 

The order updater wasn't updating payment fee when an order was updated after payment had already been added.This resulted in the order total being wrong on the checkout summary page. The fee was getting updated once the order was finalised but it wasn't added to the payment amount, this resulted in customer owing money when they paid with Stripe, the owed money being the missing payment fee. 

Because of the fix introduced by fixing #12512, the payment fees was actually correct on the summary page. 

To fix this, I made sure that payment fees are recalculated if the order has a pending payment, and then update the order total and payment total accordingly. 

##### Testing #5574 :

In the back end :
- Set up a Cash payment method with a Flat Percent calculator
- Assign it to your a shop

In the front end
- Start placing an order, selecting the Cash payment method and stop at Confirmation step
- Continue shopping by clicking on the shop name next the cart
- Add or update the quantity of an item
- Continue checkout, you should land back on the Confirmation step
-   -> check the order total and payment fees are correct
- Confirm the order
- Confirmation page and email should have the same order total and payment fee as the confirmation step

With a Stripe payment method:
- Set up a Stripe connect with a Flat Percent calculator

Use the same scenario as above

##### Testing #12512
In the back office:
- search for a completed order with a percentage fee payment method 
- add a product or update the quantity of one of the product
- check that order total and payment fee are updated accodingly

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Review by mkllnk — APPROVED** [review]
Well done for finding that! Great work.

File: spec/system/consumer/checkout/summary_spec.rb [hunk 1]
```diff
@@ -311,6 +311,47 @@
...
+        toggle_cart
+        click_link "Checkout"
+
+        # Check summary page total
+        expect(page).to have_title "Checkout Summary - Open Food Network"
+        expect(page).to have_selector("#order_total", text: 22.00)
+        expect(order.reload.payments.first.amount).to eql(22.00)
+      end
+    end
+
```
Comment (rioug) [inline spec/system/consumer/checkout/summary_spec.rb:354]: @mkllnk I was wondering if we actually need to keep this spec, this is covered by the unit test on the order updater

Comment (mkllnk) [reply spec/system/consumer/checkout/summary_spec.rb:354]: Good point. Let's remove it then.

**Review by dacook — APPROVED** [review]
Great work, well done on fixing the whole bug!

**Comment by drummer83** [comment]
Hi @rioug,

Thanks for working on this! Unfortunately I couldn't reproduce this bug in current master. I can observe the described behaviour, but it's for the *shipping fees*, not for the *payment fees*. And this actually is still the case after staging your PR.

## Before staging = current master
After editing the order and returning to checkout:
image

The calculation makes no sense at all: 50 + 5 + 5 = 60 != 58 --> order total is incorrect :x: 

Order confirmation shows the old *shipping fee*, the payment fee is updated correctly:
image

The calculation is correct, but the shipping calculation is based on the *old* cart, before adding another 20 $ product. :x:

## After staging
After editing the order and returning to checkout:
image

The calculation makes sense now: 50 + 5 + 5 = 60 --> order total is correct :heavy_check_mark: 

Order confirmation still shows the old *shipping fee*, the payment fee is updated correctly:
image

The calculation is correct, but the shipping calculation is based on the *old* cart, before adding another 20 $ product. :x:

## Conclusion
The original issue mentioned in the opening post was fixed in some earlier PR already. This PR here is about something else - but it's definitely great!
We don't have a full fix of the issue with different order totals on the order summary screen and the order confirmation screen. However, at least the calculation on the order summary page is correct now - but still the order confirmation is based on incorrect values due to the not updated shipping fees.

I'm tempted to merge this one, because it's fixing the incorrect calculation on the order summary page. However, maybe fixing the bit with the incorrect shipping fee will make this unnecessary? I'm not sure.

Adding the feedback needed label for our devs to look at it and decide.

I will also open an issue on the missing update of shipping fees. :arrow_right: #12907 

Thanks again!

**Comment by rioug** [comment]
@konrad I just tried on master, and the issue is still there.

**Comment by drummer83** [comment]
> @konrad I just tried on master, and the issue is still there.

What exactly do you mean by "the issue", @rioug?
Shopfront or backend?
Payment fee or shipping fee?

The backoffice issue regarding payment fee (opening post in #5574) I could not reproduce. That's why I closed it.

The shopfront issue regarding shipping fee is still there, but has no issue. That's why I opened #12907.

**Comment by rioug** [comment]
> What exactly do you mean by "the issue", @rioug?

I meant the payment fee not being correctly calculated in the front end,  you should be able to reproduce it in master with the testing notes for https://github.com/openfoodfoundation/openfoodnetwork/issues/5574 (see PR description)

**Comment by drummer83** [comment]
Hmm, interesting! It's a bit confusing.

You're right, in master the **Order Summary** page still uses the **old transaction fee**.
But, the **Order Confirmation** page is using the old **shipping fee**. 
The issue around the transaction fee is solved with your PR. :muscle: 

# Before staging
## Transaction fee only
### Before editing the order
image

### After editing the order
image

### Order confirmation page
image

## Transaction fee and shipping fee
### Before editing the order
image

### After editing the order
image

### Order confirmation page
image

# After staging
## Transaction fee only
### Before editing the order
image

### After editing the order
image

### Order confirmation page
image

## Transaction fee and shipping fee
### Before editing the order
image

### After editing the order
image

### Order confirmation page
image

# Conclusion
The issue around the payment fee is solved by your PR. :muscle: 
The issue around the shipping fee (#12907) is still open. This should be handled separately.
Merging this one then! :partying_face: 
Thanks again, @rioug! :pray:

**Pull request merged**